velox/exec/WindowPartition.cpp

```diff
@@ -171,4 +171,167 @@ std::pair<vector_size_t, vector_size_t> WindowPartition::computePeerBuffers(
+      firstMatch, begin, end, currentRow, orderByColumn, frameColumn);
+}
+
+template <bool isAscending>
+vector_size_t WindowPartition::linearSearchFrameValue(
+    bool firstMatch,
+    vector_size_t start,
+    vector_size_t end,
+    vector_size_t currentRow,
+    column_index_t orderByColumn,
+    column_index_t frameColumn) const {
+  auto current = partition_[currentRow];
+  bool crossedBound = false;
+  for (vector_size_t i = start; i < end; ++i) {
+    auto compareResult =
+        data_->compare(partition_[i], current, orderByColumn, frameColumn);
+
+    // The bound value was found. Return if firstMatch required.
+    // If the last match is required, then we need to find the first row that
+    // crosses the bound and return the prior row.
+    if (compareResult == 0) {
+      if (firstMatch) {
+        return i;
+      }
+    }
+
+    if constexpr (isAscending) {
+      crossedBound = compareResult > 0;
+    } else {
+      crossedBound = compareResult < 0;
+    }
+    // Bound is crossed. Last match needs the previous row.
+    // But for first row matches, this is the first
+    // row that has crossed, but not equals boundary (The equal boundary case
+    // is covered by the condition above). So the bound matches this row itself.
+    if (crossedBound) {
+      if (firstMatch) {
+        return i;
+      } else {
+        return i - 1;
+      }
+    }
+  }
+
+  // Return a row beyond the partition boundary. The logic to determine valid
+  // frames handles the out of bound and empty frames from this value.
+  return end == numRows() ? numRows() + 1 : -1;
+}
+
+template <bool isAscending>
+void WindowPartition::updateKRangeFrameBounds(
+    bool firstMatch,
+    bool isPreceding,
+    vector_size_t startRow,
+    vector_size_t numRows,
+    column_index_t frameColumn,
+    const vector_size_t* rawPeerBounds,
+    vector_size_t* rawFrameBounds) const {
+  column_index_t orderByColumn = sortKeyInfo_[0].first;
+  RowColumn frameRowColumn = data_->columnAt(frameColumn);
+
+  vector_size_t start = 0;
+  vector_size_t end = partition_.size();
+  for (auto i = 0; i < numRows; i++) {
+    auto currentRow = startRow + i;
+    bool frameIsNull = RowContainer::isNullAt(
+        partition_[currentRow],
+        frameRowColumn.nullByte(),
+        frameRowColumn.nullMask());
+
+    // For NULL values, CURRENT ROW semantics apply. So get frame bound from
+    // peer buffer.
+    if (frameIsNull) {
+      rawFrameBounds[i] = rawPeerBounds[i];
+    } else {
+      // If the search is for a preceding bound then rows between
+      // [0, currentRow] are examined. For following bounds, rows between
+      // [currentRow, numRows()) are checked.
+      if (isPreceding) {
+        start = 0;
+        end = currentRow + 1;
+      } else {
+        start = currentRow;
+        end = partition_.size();
+      }
+      rawFrameBounds[i] = searchFrameValue<isAscending>(
+          firstMatch, start, end, currentRow, orderByColumn, frameColumn);
```

Comment:
Hi @aditi-pandit, can this part be optimized by narrowing down the search range dynamically?
For start bound search, considering we are doing the search on ascending ordered column, I think the current row's search range can be `[lastSearchResult, currentRow]`, instead of `[0, currentRow]`.

See below example for start bound search (1 preceding). From last search, we get the start bound for 6 is 5 (rowID  = 2). Then, when searching start bound for 7, as `current_row >= last_row` is always true for ascending ordered column, we know `current_row - 1 > = last_row - 1`. Thus, we can start the search from rowID = 2 (last search result) instead of 0.

```
rowID     column (ordered)
0                  3
1                  4
2                  5
3                  6
4                  7
...                 ...
```

The end bound search may also be optimized similarly. This optimized search range looks like a sliding window. Does it make sense?

Reply:
@PHILO-HE : Its a good idea to optimize this search using the previously row frame value. I will explore that in a subsequent PR.

Reply:
@aditi-pandit Aditi, would you create a GitHub issue about this optimization?

Reply:
Have created https://github.com/facebookincubator/velox/issues/7564.